pandas/core/indexes/base.py

```diff
@@ -212,6 +212,11 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
                     if is_integer_dtype(dtype):
                         inferred = lib.infer_dtype(data)
                         if inferred == 'integer':
+                            if is_unsigned_integer_dtype(dtype) and\
```

Comment:
move this logic to pandas.core.dtypes.cast (everything below is_integer_dtype, e.g. 213-220)

```
def maybe_cast_to_integer(arr, dtype):
....
```

this would give you an integer or unsigned integer array (or raise if the dtype is incompat)